### PR TITLE
docs: addition of param and result sections

### PR DESCRIPTION
**Pull request by hachreak**
- Improves documentation. (addresses #57)

Signed-off-by: Leonardo Rossi leonardo.r@cern.ch

### CODE REVIEW

File: invenio_oauth2server/ext.py [hunk 1]
```diff
@@ -76,21 +76,32 @@ def __init__(self, app, entry_point_group=None):
             self.load_entry_point_group(entry_point_group)
 
     def scope_choices(self, exclude_internal=True):
-        """Return list of scope choices."""
+        """Return list of scope choices.
+
+        :param exclude_internal: Exclude or not internal scopes
```
Comment (jirikuncar) [inline invenio_oauth2server/ext.py]: scopes.

File: invenio_oauth2server/ext.py [hunk 2]
```diff
@@ -76,21 +76,32 @@ def __init__(self, app, entry_point_group=None):
...
+            (Default: ``True``).
```
Comment (jirikuncar) [inline invenio_oauth2server/ext.py]: (Default: ...)

File: invenio_oauth2server/models.py [hunk 3]
```diff
@@ -362,19 +362,28 @@ class Token(db.Model):
...
+        :param scopes: The list of scopes.
+        """
         validate_scopes(scopes)
         self._scopes = " ".join(set(scopes)) if scopes else ""
 
     def get_visible_scopes(self):
-        """Get list of non-internal scopes for token."""
+        """Get list of non-internal scopes for token.
+
+        :returnsa: A list of scopes.
```
Comment (jirikuncar) [inline invenio_oauth2server/models.py]: returnsa -> returns

**Comment by jirikuncar** [comment]
@hachreak 37126f0 does not close `#59` since we still don't have the new documentation structure.

**Comment by hachreak** [comment]
@alizeepace could you have a look also in this PR? thanks :D

File: invenio_oauth2server/ext.py [hunk 4]
```diff
@@ -76,21 +76,32 @@ def __init__(self, app, entry_point_group=None):
             self.load_entry_point_group(entry_point_group)
 
     def scope_choices(self, exclude_internal=True):
-        """Return list of scope choices."""
+        """Return list of scope choices.
+
+        :param exclude_internal: Exclude or not internal scopes.
```
Comment (alizeepace) [inline invenio_oauth2server/ext.py]: Exclude internal scopes or not.

File: invenio_oauth2server/ext.py [hunk 5]
```diff
@@ -76,21 +76,32 @@ def __init__(self, app, entry_point_group=None):
...
+            (Default: ``True``)
+        :returns: A list of tuple (id, scope).
```
Comment (alizeepace) [inline invenio_oauth2server/ext.py]: A list of tuples

File: invenio_oauth2server/validators.py [hunk 6]
```diff
@@ -51,7 +53,13 @@ def validate_redirect_uri(value):
 
 
 def validate_scopes(value_list):
-    """Validate if each element in a list is a registered scope."""
+    """Validate if each element in a list is a registered scope.
+
+    :param value_list: The list of scopes.
+    :raises invenio_oauth2server.errors.ScopeDoesNotExists: The exception is
+        raised if a scope is not registered.
+    :returns: ``True`` if it's successfully validate.
```
Comment (alizeepace) [inline invenio_oauth2server/validators.py]: validated

**Comment by hachreak** [comment]
@jirikuncar any idea why?
```
Warning, treated as error:

/home/travis/build/inveniosoftware/invenio-oauth2server/docs/api.rst:59: WARNING: unknown document: i18n
```

File: invenio_oauth2server/views/settings.py [hunk 7]
```diff
@@ -70,7 +70,11 @@ def decorated(*args, **kwargs):
 
 
 def token_getter(is_personal=True, is_internal=False):
-    """Decorator to retrieve Token object and check user permission."""
+    """Decorator to retrieve Token object and check user permission.
+
+    :param is_personal: Search for a personal token. (Default: ``True``)
+    :param is_internal: Search for a internal token. (Default: ``True``)
```
Comment (jirikuncar) [inline invenio_oauth2server/views/settings.py]: default is False

File: invenio_oauth2server/provider.py [hunk 8]
```diff
@@ -38,13 +38,17 @@
...
 
     Needed for grant type 'password'. Note, grant type password is by default
     disabled.
+
+    :param email: User email.
+    :param password: Password.
+    :returns: The user instance or ``None``.
     """
-    user = User.query.filter_by(username=username).first()
+    user = User.query.filter_by(email=email).first()
```
Comment (jirikuncar) [inline invenio_oauth2server/provider.py]: WDYT about following improvement?

```python
user = User.query.filter_by(email=email).one_or_none()
if user and user.check_password(password):
    return user
```

Do we have a test that tries to fetch non-existent user?

Comment (hachreak) [inline invenio_oauth2server/provider.py]: maybe it's better to use `datastore` and `flask-security` `verify_password`?
```python
from flask_security.utils import verify_password
datastore = LocalProxy(lambda: current_app.extensions['security'].datastore)
user = datastore.find_user(email=email)
if user and verify_password(password, user.password):
    return user
```

File: invenio_oauth2server/provider.py [hunk 9]
```diff
@@ -29,31 +29,45 @@
...
 
-@oauth2.usergetter
-def get_user(username, password, *args, **kwargs):
+def _get_user(email, password, *args, **kwargs):
     """Get user for grant type password.
 
     Needed for grant type 'password'. Note, grant type password is by default
     disabled.
+
+    :param email: User email.
```
Comment (jirikuncar) [inline invenio_oauth2server/provider.py]: IMHO we don't need to document obvious arguments in non-api function.

File: invenio_oauth2server/provider.py [hunk 10]
```diff
@@ -29,31 +29,45 @@
...
+    :returns: The user instance or ``None``.
     """
-    user = User.query.filter_by(username=username).first()
-    if user.check_password(password):
+    user = datastore.find_user(email=email)
+    if user and verify_password(password, user.password):
         return user
 
 
+get_user = oauth2.usergetter(_get_user)
```
Comment (jirikuncar) [inline invenio_oauth2server/provider.py]: Why can't you use the decorator?

File: tests/test_provider.py [hunk 11]
```diff
@@ -33,6 +33,19 @@
...
 
 from invenio_oauth2server.models import Client, Token
+from invenio_oauth2server.provider import _get_user
+
+
+def test_get_user(provider_fixture):
+    """Test get user."""
+    app = provider_fixture
+    with app.app_context():
+        assert _get_user(email='user@not.found', password='password') is None
```
Comment (jirikuncar) [inline tests/test_provider.py]: We should test high-level `oauth2.usergetter` method in action and not the "low-level" implementation directly.

**Comment by jirikuncar** [comment]
@hachreak can you please update the PR?

File: invenio_oauth2server/provider.py (hunk 9 above)
Comment (jirikuncar) [inline invenio_oauth2server/provider.py]: ❓

File: invenio_oauth2server/provider.py (hunk 10 above)
Comment (jirikuncar) [inline invenio_oauth2server/provider.py]: ❓

File: setup.py [hunk 12]
```diff
@@ -79,6 +79,9 @@
     'Flask-Breadcrumbs>=0.3.0',
     'Flask-Login>=0.3.0',
     'Flask-OAuthlib>=0.9.3',
+    # FIXME oauthlib v2 has some incompatibility with Flask-OAuthlib 0.9.3.
+    # see #78
+    'oauthlib>=1.1.2,<2.0.0',
```
Comment (jirikuncar) [inline setup.py]: see line 92

**Comment by jirikuncar** [comment]
@hachreak please remove the last commit or allow changes by maintainers. Thank you

**Comment by jirikuncar** [comment]
LGTM

**Pull request merged**